### PR TITLE
rocAL - Pybind changes for tensor augmentations

### PR DESCRIPTION
**Pull request by SundarRajan28**
- Add pybind support for all node augmentations
- Modified the python unit tests to dump and compare augmentation outputs to the golden outputs similar to C++ unittests
- Added changes to generic iterator to support multiple augmentation outputs

Note: The previous PRs (#31 , #32 , #33 ) has to be merged before this PR for building and checking the pybind and python unit tests in this PR

### CODE REVIEW

File: rocAL/rocAL_pybind/amd/rocal/fn.py [hunk 1]
```diff
@@ -29,219 +29,437 @@
 import rocal_pybind as b
 from amd.rocal.pipeline import Pipeline
 
-def blend(*inputs,ratio=None):
-    kwargs_pybind = {"input_image0":inputs[0], "input_image1":inputs[1], "is_output":False ,"ratio":ratio}
-    blend_image = b.Blend(Pipeline._current_pipeline._handle ,*(kwargs_pybind.values()))
+
+def blend(*inputs, ratio=None, rocal_tensor_output_layout=types.NHWC, rocal_tensor_output_datatype=types.UINT8):
+    """
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/amd/rocal/fn.py:34]: @SundarRajan28 we need to confirm with Kiriti how to pass API descriptions for Python in Doxygen so that it auto generates.

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 1 above)
Comment (LakshmiKumar23) [reply rocAL/rocAL_pybind/amd/rocal/fn.py:34]: @SundarRajan28 PLease follow this example and make changes accordingly: https://github.com/Feneric/doxypypy#example

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 2]
```diff
@@ -22,6 +22,7 @@
 import amd.rocal.types as types
 import numpy as np
 import cupy as cp
+import torch
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py]: @SundarRajan28 why  are we brining this back? This would make  torch as a pre req for all use cases. That's why we moved to cupy

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 3]
```diff
@@ -157,39 +158,36 @@ def build(self):
...
+            elif (isinstance(array, torch.Tensor)):
+                return b.getOneHotEncodedLabels(self._handle, ctypes.c_void_p(array.data_ptr()), self._num_classes, 0)
         else:
-            if (isinstance(array,cp.ndarray)):
-                b.getCupyOneHotEncodedLabels(self._handle, array.data.ptr, self._numOfClasses, 1)
-            else: #torch tensor
-                return b.getOneHotEncodedLabels(self._handle, ctypes.c_void_p(array.data_ptr()), self._numOfClasses, 1)
+            if (isinstance(array, cp.ndarray)):
+                b.getCupyOneHotEncodedLabels(self._handle, array.data.ptr, self._num_classes, 1)
+            elif (isinstance(array, torch.Tensor)):
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py]: @SundarRajan28 same comment as above

File: rocAL/rocAL_pybind/amd/rocal/types.py [hunk 4]
```diff
@@ -18,71 +18,7 @@
...
-from rocal_pybind.types import SCALING_MODE_NOT_LARGER
-
-#     RocalResizeInterpolationType
-from rocal_pybind.types import NEAREST_NEIGHBOR_INTERPOLATION
-from rocal_pybind.types import LINEAR_INTERPOLATION
-from rocal_pybind.types import CUBIC_INTERPOLATION
-from rocal_pybind.types import LANCZOS_INTERPOLATION
-from rocal_pybind.types import GAUSSIAN_INTERPOLATION
-from rocal_pybind.types import TRIANGULAR_INTERPOLATION
+from rocal_pybind.types import *
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/amd/rocal/types.py]: @SundarRajan28 * import will give codacy errors. Please revert back

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py [hunk 5]
```diff
@@ -52,154 +73,290 @@ def main():
...
+        for i, (output_list, labels) in enumerate(data_loader, 0):
+            for j in range(len(output_list)):
+                if args.print_tensor:
+                    print("**************", i, "*******************")
+                    print("**************starts*******************")
+                    print("\nImages:\n", output_list[j])
+                    print("\nLABELS:\n", labels)
+                    print("**************ends*******************")
+                    print("**************", i, "*******************")
+                draw_patches(output_list[j], cnt, rocal_device)
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py:355]: @SundarRajan28 are we changing this to show the image for all iterations? it used to only for the 1st?

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 2 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py]: Removed the torch import and usage of torch APIs in the latest commit

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 3 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py]: Removed the torch import and usage of torch APIs in the latest commit

File: rocAL/rocAL_pybind/amd/rocal/types.py (hunk 4 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/types.py]: Removed the * import and reverted back to the manual imports

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py (hunk 5 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py:355]: We modified the python unit tests to replicate the same functionality as C++ unit tests which dumps all the iteration outputs with the same filename. The last iteration outputs is taken as the augmentation output to be compared against the golden outputs

File: rocAL/rocAL_pybind/amd/rocal/decoders.py [hunk 6]
```diff
@@ -107,7 +107,7 @@ def image(*inputs, user_feature_key_map=None, path='', file_root='', annotations
             "max_width": max_decoded_width,
             "max_height": max_decoded_height,
             "dec_type": decoder_type}
-        decoded_image = b.ImageDecoderShard(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
+        decoded_image = b.imageDecoderShard(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/decoders.py:110]: you need to make sure all the tests on python are changed accordingly because of the name change in api

File: rocAL/rocAL_pybind/amd/rocal/fn.py [hunk 7]
```diff
@@ -29,219 +29,437 @@
...
-    fog_image = b.Fog(Pipeline._current_pipeline._handle ,*(kwargs_pybind.values()))
+    kwargs_pybind = {"input_image": inputs[0],
+                     "is_output": False, "fog_value": fog, "rocal_tensor_output_layout": rocal_tensor_output_layout, "rocal_tensor_output_datatype": rocal_tensor_output_datatype}
+    fog_image = b.fog(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
     return (fog_image)
 
-def brightness(*inputs, brightness=1.0, bytes_per_sample_hint=0, image_type=0,
-               preserve=False, seed=-1, device=None):
+
+def brightness(*inputs, alpha=None, beta=None, rocal_tensor_output_layout=types.NHWC, rocal_tensor_output_datatype=types.UINT8):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/fn.py]: I suggest retaining the brightness name for the parameter. What is the reasoning for name change?

File: rocAL/rocAL_pybind/amd/rocal/fn.py [hunk 8]
```diff
@@ -255,339 +473,536 @@ def resize(*inputs, bytes_per_sample_hint=0, image_type=0, interp_type=1, mag_fi
...
-    hue = b.CreateFloatParameter(hue) if isinstance(hue, float) else hue
-    saturation = b.CreateFloatParameter(saturation) if isinstance(
-        saturation, float) else saturation
-    # pybind call arguments
-    kwargs_pybind = {"input_image0": inputs[0], "output_layout": output_layout, "output_dtype": output_dtype, "is_output": False,
-                     "alpha": brightness, "beta": contrast, "hue": hue, "sat": saturation}
-    color_twist_image = b.ColorTwist(Pipeline._current_pipeline._handle ,*(kwargs_pybind.values()))
+
+def color_twist(*inputs, brightness=1.0, contrast=1.0, hue=0.0,
+                saturation=1.0, rocal_tensor_output_layout=types.NHWC, rocal_tensor_output_datatype=types.UINT8):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/fn.py]: why do we changes the name to rocal_tensor_xxx and rocal_tensor_output_xx etc. The previous name is more appropriates

File: rocAL/rocAL_pybind/amd/rocal/fn.py [hunk 9]
```diff
@@ -255,339 +473,536 @@ def resize(*inputs, bytes_per_sample_hint=0, image_type=0, interp_type=1, mag_fi
...
+    # pybind call arguments
+    kwargs_pybind = {"input_image": inputs[0], "is_output": False, "p_alpha": brightness, "p_beta": contrast,
+                     "p_hue": hue, "p_sat": saturation, "rocal_tensor_output_layout": rocal_tensor_output_layout, "rocal_tensor_output_datatype": rocal_tensor_output_datatype}
+    color_twist_image = b.colorTwist(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
     return (color_twist_image)
 
-def uniform(*inputs,rng_range=[-1, 1], device=None):
-    output_param = b.CreateFloatUniformRand(rng_range[0], rng_range[1])
+
+def uniform(*inputs, rng_range=[-1, 1]):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/fn.py]: we need to add the device parameter back. true for all the operators

File: rocAL/rocAL_pybind/amd/rocal/fn.py [hunk 10]
```diff
@@ -255,339 +473,536 @@ def resize(*inputs, bytes_per_sample_hint=0, image_type=0, interp_type=1, mag_fi
...
+    rng_range (list of ints, optional, default = [-1, 1]) - uniform distribution used for random number generation
+    """
+    output_param = b.createFloatUniformRand(rng_range[0], rng_range[1])
     return output_param
 
-def random_bbox_crop(*inputs,all_boxes_above_threshold=True, allow_no_crop=True, aspect_ratio=None, bbox_layout="", bytes_per_sample_hint=0,
-                crop_shape=None, input_shape=None, ltrb=True, num_attempts=1 ,scaling=None, preserve=False, seed=1, shape_layout="",
-                threshold_type="iou", thresholds=None, total_num_attempts=0, device=None, labels=None):
+
+def random_bbox_crop(*inputs, all_boxes_above_threshold=True, allow_no_crop=True, aspect_ratio=None,
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/fn.py]: what happened to the other parameters line ltrb, threshold_type etc. We need to keep all relevant parameters name same as dali for easy porting of dali to rocal

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 11]
```diff
@@ -140,8 +140,8 @@ def __init__(self, batch_size=-1, num_threads=0, device_id=-1, seed=1,
         self._BoxEncoder = None
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py]: please change all variables to use snake_case

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 12]
```diff
@@ -198,82 +195,82 @@ def __enter__(self):
...
-    def GetImageId(self, array):
+    def get_image_id(self, array):
         b.getImageId(self._handle, array)
 
-    def GetBoundingBoxCount(self):
+    def get_bounding_box_count(self):
         return b.getBoundingBoxCount(self._handle)
 
-    def GetBoundingBoxLabels(self):
+    def get_bounding_box_abels(self):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py]: I thought we will be returning tensor lists for these APIs.

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py [hunk 13]
```diff
@@ -52,154 +73,290 @@ def main():
...
     # Create Pipeline instance
-    pipe = Pipeline(batch_size=batch_size, num_threads=num_threads, device_id=local_rank, seed=random_seed, rocal_cpu=rocal_cpu, tensor_layout=types.NHWC if args.NHWC else types.NCHW , tensor_dtype=types.FLOAT16 if args.fp16 else types.FLOAT)
+    pipe = Pipeline(batch_size=batch_size,
+                    num_threads=num_threads,
+                    device_id=local_rank,
+                    seed=random_seed,
+                    rocal_cpu=rocal_cpu,
+                    tensor_layout=tensor_layout,
+                    tensor_dtype=tensor_dtype,
+                    output_memory_type=types.CPU_MEMORY if rocal_cpu else types.GPU_MEMORY)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py]: why do we need to pass out_memory_type here. Shouldn't it be implied?
Also I would suggest using HOST/device instead of CPU_MEMORY and GPU_MEMORY

File: rocAL/rocAL_pybind/examples/PYTHON_UNITTEST_TEST_FILE.sh [hunk 14]
```diff
@@ -1,149 +1,118 @@
...
-
-        if [ -z "$print_tensor" ];
-    then
-        print_tensor_arg=print_tensor #True by default
+for ((device=dev_start;device<=dev_end;device++))
+do 
+    if [ $device -eq 1 ]
+    then 
+        device_name="hip"
+        backend_arg=rocal-gpu
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/examples/PYTHON_UNITTEST_TEST_FILE.sh:69]: @fiona-gladwin Has this been tested on both docker and bare metal for both backends. Please talk to Swetha, I had a call with her regarding this

File: rocAL/rocAL_pybind/examples/image_comparison.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from PIL import Image
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/examples/image_comparison.py:21]: @fiona-gladwin please add the requirement for ```pip install Pillow``` in the readme

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py [hunk 16]
```diff
@@ -52,154 +73,290 @@ def main():
...
+    if (scaling_mode != types.SCALING_MODE_DEFAULT and interpolation_type !=
+            types.LINEAR_INTERPOLATION):
+        interpolation_type = types.LINEAR_INTERPOLATION
+    if augmentation_name in ["hue", "saturation", "color_twist"] and color_format == types.GRAY:
+        print("Not a valid option! Exiting!")
+        sys.exit(0)
+
     try:
-        path= "OUTPUT_IMAGES_PYTHON/NEW_API/FILE_READER/" + args.augmentation_name
+        path = "OUTPUT_IMAGES_PYTHON/NEW_API/FILE_READER/" + args.augmentation_name
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py]: @fiona-gladwin I have mentioned this before. Please remove New_api from all tests. It doesn't make sense to anyone outside

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py [hunk 17]
```diff
@@ -52,154 +73,290 @@ def main():
...
+                    seed=random_seed,
+                    rocal_cpu=rocal_cpu,
+                    tensor_layout=tensor_layout,
+                    tensor_dtype=tensor_dtype,
+                    output_memory_type=types.CPU_MEMORY if rocal_cpu else types.GPU_MEMORY)
     # Set Params
     output_set = 0
     rocal_device = 'cpu' if rocal_cpu else 'gpu'
-    #hardcoding decoder_device to cpu until VCN can decode all JPEGs
+    # hardcoding decoder_device to cpu until VCN can decode all JPEGs
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py]: @fiona-gladwin Is this still the case? With rocm5.6 all JPEGs should be enabled on Navi and MI cards

File: rocAL/rocAL_pybind/rocal_pybind.cpp [hunk 18]
```diff
@@ -666,111 +482,119 @@ namespace rocal{
...
+            py::return_value_policy::reference);
+        m.def("videoDecoderResize", &rocalVideoFileResize, "Reads videos from the source given and decodes it according to the policy only for videos as inputs. Resizes the decoded frames to the dest width and height.",
+            py::return_value_policy::reference);
+        m.def("sequenceReader", &rocalSequenceReader, "Creates JPEG image reader and decoder. Reads [Frames] sequences from a directory representing a collection of streams.",
+            py::return_value_policy::reference);
+        m.def("rocalResetLoaders", &rocalResetLoaders);
+        m.def("videoMetaDataReader", &rocalCreateVideoLabelReader, py::return_value_policy::reference);
         // rocal_api_augmentation.h
-        m.def("SSDRandomCrop",&rocalSSDRandomCrop,
+        // m.def("SSDRandomCrop", &rocalSSDRandomCrop,
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/rocal_pybind.cpp]: @fiona-gladwin why is this commented out? I see SSD Random crop was added in fn.py

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 7 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py]: In ToT, the brightness parameter in this function is unused and we pass alpha and beta as None to C++ regardless of value passed. So this function is changed to fetch the alpha and beta values from the user instead and pass it to C++

File: rocAL/rocAL_pybind/amd/rocal/decoders.py (hunk 6 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/decoders.py:110]: We have changed all the API names accordingly. For readers.py and decoders.py, we only changed the API names for the file reader and image decoder. Other readers and decoders are changed in the next metanodes PR and the example files are changed in the python examples PR

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 12 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py]: This pipeline function calls the getBoundingBoxLabels function in rocal_pybind.cpp which returns the labels as lists.

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py (hunk 13 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py]: The copy_data function in tensor.cpp requires the memory type passed as the second argument. If its not passed, it sets memory type as CPU_MEMORY and assumes the ptr passed is a host buffer ptr and uses memcpy to transfer data. To avoid this, we set the memory type in pipeline and pass the memory type to the copy_data function in C++

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 10 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py]: I retained all the parameters that are passed to C++ but some of the other parameters were present only to make API parameters similar to DALI but were not passed to C++. I removed those unused parameters from the function

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py (hunk 17 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py]: Currently we are comparing the augmentation outputs with golden outputs taken with turbojpeg decoder so hardcoding to cpu for now. We will support both backends once the respective golden outputs are taken. I'll remove the "until VCN can decode all jpegs" part of the comment

File: rocAL/rocAL_pybind/examples/PYTHON_UNITTEST_TEST_FILE.sh (hunk 14 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/PYTHON_UNITTEST_TEST_FILE.sh:69]: I have tested the python unit tests on both docker and bare metal - its working as expected

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py (hunk 16 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py]: Removed the folder name from the python unit tests

File: rocAL/rocAL_pybind/rocal_pybind.cpp (hunk 18 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/rocal_pybind.cpp]: Uncommented the SSD random crop augmentation in pybind

File: rocAL/rocAL_pybind/examples/image_comparison.py (hunk 15 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/image_comparison.py:21]: Added pip install pillow to the README in rocal_pybind folder

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 9 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py]: Added device parameter to all functions in fn.py

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 8 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py]: Changed the names to output_layout and output_dtype

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 7 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py]: I have renamed the brightness parameters similar to DALI. I also renamed all parameters from other functions to their DALI counterparts

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 10 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py]: As discussed in last call, added most of the parameters from DALI even if its unused and will be adding docstrings for those params as "Parameter unused in augmentation" in the doxygen PR

File: rocAL/rocAL_pybind/README.md [hunk 19]
```diff
@@ -20,7 +20,7 @@ rocAL_pybind installs during [MIVisionX build](https://github.com/GPUOpen-Profes
 
 * Install pip packages
 ````
-pip3 install numpy opencv-python torch
+pip3 install numpy opencv-python torch pillow
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/README.md:23]: @fiona-gladwin Seems like capital letter 'P'. Did you try this?
https://pypi.org/project/Pillow/

Comment (SundarRajan28) [reply rocAL/rocAL_pybind/README.md:23]: Both pillow and Pillow installs the same pip package

**Review by LakshmiKumar23 — APPROVED** [review]

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py (hunk 13 above)
Comment (LakshmiKumar23) [reply rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py]: @SundarRajan28 Please use host and device as mentioned

File: rocAL/rocAL_pybind/amd/rocal/fn.py [hunk 20]
```diff
@@ -255,339 +513,578 @@ def resize(*inputs, bytes_per_sample_hint=0, image_type=0, interp_type=1, mag_fi
...
+    num_attempts (int, optional, default = 1) - number of attempts to get a crop window that matches the area factor and aspect ratio conditions
+
+    device (string, optional, default = None) - Backend used for the augmentation - either 'cpu' or 'gpu'
+    
+    output_layout (int, optional, default = types.NHWC) - tensor layout for the augmentation output
+
+    output_dtype (int, optional, default = types.UINT8) - tensor dtype for the augmentation output
     """
-    if(num_attempts == 1):
+    if (num_attempts == 1):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/fn.py:654]: Is this correct?

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 10 above)
Comment (rrawther) [reply rocAL/rocAL_pybind/amd/rocal/fn.py]: We need ltrb and threashold type. Does the low level api assume it is ltrb?

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 21]
```diff
@@ -410,12 +405,12 @@ def create_pipeline(*args, **kwargs):
             with pipe:
                 pipe_outputs = func(*args, **fn_kwargs)
                 if isinstance(pipe_outputs, tuple):
-                    po = pipe_outputs
+                    outputs = pipe_outputs
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py]: why is this renamed. po (pipe_outputs) was fine

File: rocAL/rocAL_pybind/amd/rocal/plugin/generic.py [hunk 22]
```diff
@@ -22,66 +22,78 @@
 import numpy as np
 import rocal_pybind as b
 import amd.rocal.types as types
-import ctypes
+
 
 class ROCALGenericIterator(object):
-    def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = False, multiplier = [1.0,1.0,1.0], offset = [0.0, 0.0, 0.0], tensor_dtype=types.FLOAT, display=False, device="cpu", device_id =0):
+    def __init__(self, pipeline, tensor_layout=types.NCHW, reverse_channels=False, multiplier=[1.0, 1.0, 1.0], offset=[0.0, 0.0, 0.0], tensor_dtype=types.FLOAT, display=False, device="cpu", device_id=0):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/plugin/generic.py:28]: This line change just removed all the spacing for default arguments. I think having spacing like before is better and keep the same coding style everywhere

File: rocAL/rocAL_pybind/examples/PYTHON_UNITTEST_TEST_FILE.sh (hunk 14 above)
Comment (rrawther) [reply rocAL/rocAL_pybind/examples/PYTHON_UNITTEST_TEST_FILE.sh:69]: Lakshmi, Please check if the issue with the script still exist or not

File: rocAL/rocAL_pybind/rocal_pybind.cpp [hunk 23]
```diff
@@ -72,226 +72,26 @@ namespace rocal{
...
+    py::object wrapper_image_name(RocalContext context, int array_len)
     {
         py::array_t<char> array;
         auto buf = array.request();
-        char* ptr = (char*) buf.ptr;
-        ptr = (char *)calloc(array_len, sizeof(char));
+        char* ptr = static_cast<char *>(buf.ptr);
+        ptr = static_cast<char *>(calloc(array_len, sizeof(char)));
         // call pure C++ function
         rocalGetImageName(context,ptr);
```
Comment (rrawther) [inline rocAL/rocAL_pybind/rocal_pybind.cpp:88]: add space between arguments and follow same coding style throughout

File: rocAL/rocAL_pybind/rocal_pybind.cpp [hunk 24]
```diff
@@ -666,111 +482,119 @@ namespace rocal{
...
-            py::return_value_policy::reference);
-        m.def("FusedDecoderCropShard",&rocalFusedJpegCropSingleShard,"Reads file from the source and decodes them partially to output random crops",
-            py::return_value_policy::reference);
-        m.def("TF_ImageDecoderRaw",&rocalRawTFRecordSource,"Reads file from the source given and decodes it according to the policy only for TFRecords",
-              py::return_value_policy::reference);
-        m.def("Cifar10Decoder",&rocalRawCIFAR10Source,"Reads file from the source given and decodes it according to the policy only for TFRecords",
-              py::return_value_policy::reference);
-        m.def("VideoDecoder",&rocalVideoFileSource,"Reads videos from the source given and decodes it according to the policy only for Videos as inputs",
+        // Will be enabled when the other decoders are implemented in C++
+        // m.def("COCO_ImageDecoderSlice", &rocalJpegCOCOFileSourcePartial,"Reads file from the source given and decodes it according to the policy",
```
Comment (rrawther) [inline rocAL/rocAL_pybind/rocal_pybind.cpp:486]: why are all these commented

Comment (SundarRajan28) [reply rocAL/rocAL_pybind/rocal_pybind.cpp:486]: Readers and decoders other than file reader and image decoder are not implemented in this PR so uncommenting this will lead to build errors

File: rocAL/rocAL_pybind/amd/rocal/plugin/generic.py (hunk 22 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/plugin/generic.py:28]: We checked with google coding style guidelines for python and changed the spacing accordingly.
https://google.github.io/styleguide/pyguide.html#3194-default-values

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 20 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py:654]: We checked with google coding style guidelines for python and changed the spacing accordingly. 
https://google.github.io/styleguide/pyguide.html

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py (hunk 13 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py]: Renamed both types as HOST_MEMORY and DEVICE_MEMORY respectively

**Review by rrawther — CHANGES_REQUESTED** [review]
pl address review comments

File: rocAL/rocAL_pybind/rocal_pybind.cpp (hunk 23 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/rocal_pybind.cpp:88]: We are using the same coding style for all static_casts in C++ source files as well.

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 21 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py]: Reverted back to the original name for the variable

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 10 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py]: I've added all the parameters for random_bbox_crop similar to DALI in the latest commit

**Comment by SundarRajan28** [comment]
@rrawther I've resolved the review comments and added changes to the PR

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 20 above)
Comment (rrawther) [reply rocAL/rocAL_pybind/amd/rocal/fn.py:654]: I was referring to assigning _num_attempts = 20 if it is set 1?

**Review by rrawther — APPROVED** [review]
make sure this is tested

File: rocAL/rocAL_pybind/amd/rocal/fn.py (hunk 20 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/fn.py:654]: DALI has default value of 1 for this param so we added the same but it is set as 20 if the default value is used for SSD random crop.
